### PR TITLE
Added Camera terrain collision and camera restriction flag in scene

### PR DESCRIPTION
**Pull request by The-Spaghetti-Is-Ready**
## What problem does this pull request address?

This change adds camera terrain collision back after restructuring the entrypoint so that `Scene` is a standard interface. It had to be removed in that refactor since it was based on a global terrain object. Now it is fully dynamic.

## How does this pull request solve the problem?

This change is implemented by adding a function into `systems` interface which checks for collision with any terrain components that are added into a scene. If there is a value present in any of terrain components which match the current `x` and `z` coordinates of the camera, the camera will be set to the resulting height of that terrain object. There is also an additional optional camera offset that is passed into this function to adjust player/camera height (currently set to `10.0f` by default).

The `scene` interface now contains a `restrictCamera` bool flag which determines whether or not terrain collision is active. This flag is currently set to change if `r` is pressed by the user when running the current scene through the static key callback function.

The concern with calculating terrain collisions in this method is that adding a lot of terrain entities into the scene may prove to negatively affect performance as currently the system has to iterate through all terrain components present to check that camera is within its bounds. Fortunately we are not required to load more than one terrain so performance will not be affected by the specific game implementation for this project.

The camera height translation is based on the model matrices given to the Terrain instances from the components in their scripts. This is done since this transform component is applied to the Terrain when its constructed, and using the model matrix based on the transform component allows for camera collision to be based on the true, real-time translations of the terrain since the transform component values get updated by the systems, and not the raw terrain data itself.

The `updateTransformComponents` function was removed from `Systems` in this PR as well as that function is not being used anywhere and is not necessary.

The Terrain texture was also swapped out in this PR to make the terrain look more visually appealing. Terrain was also scaled up to better suit assignment terrain scale requirements. 

https://github.com/Minywire/Bottlebrush_Engine/assets/77040582/5bf9aa94-29a3-42b0-8ee1-abbc46bffc41

## What testing has been performed?

- [x] Builds successfully
- [x] Runs successfully
- [x] Camera accurately follow terrain and doesn't jitter when standing still
- [x] Stable and high framerate

### CODE REVIEW

File: Bottlebrush_Engine/Bottlebrush_Engine/BBComponentSystem/Systems.cpp [hunk 1]
```diff
@@ -283,4 +267,38 @@ void Systems::updateAI(ECS& ecs, sol::state& lua_state, float deltaTime) {
...
+        model = glm::scale(model, terrainTransform.scale);
+
+        const glm::vec3 local_position =
+            glm::inverse(model) * glm::vec4(camera.GetPosition(), 1);
+
+        const std::optional<float> terrain_height =
+            currentTerrain.GetHeight(local_position.x, local_position.z);
+
+        if (terrain_height.has_value())
+        {
```
Comment (Idraliski) [inline Bottlebrush_Engine/Bottlebrush_Engine/BBComponentSystem/Systems.cpp:296]: Nitpicking: not sure if this is your style, or if you plan on adding more to this. But I'd like the below, so its less lines
`terrain_height.has_value() ? camera.SetPositionY(terrain_height.value() + offset_y) : camera.SetPositionY(0);`
it just avoids the if else statement if you only plan on doing a single action for true and false.

**Review by Idraliski — APPROVED** [review]
LGTM, I have 1 nitpicking comment. But feel free to ignore it

File: Bottlebrush_Engine/Bottlebrush_Engine/BBComponentSystem/Systems.cpp (hunk 1 above)
Comment (The-Spaghetti-Is-Ready) [reply Bottlebrush_Engine/Bottlebrush_Engine/BBComponentSystem/Systems.cpp:286]: Got rid of the setting camera position to 0 so it's fine now

File: Bottlebrush_Engine/Bottlebrush_Engine/BBScene/Scene.cpp [hunk 2]
```diff
@@ -157,6 +159,7 @@ void Scene::init()
     glfwSetKeyCallback(window.GetContext(), KeyCallback);
     glfwSetCursorPosCallback(window.GetContext(), MouseCallback);
     glfwSetScrollCallback(window.GetContext(), ScrollCallback);
+    glfwWindowHint(GLFW_DEPTH_BITS, 32);  // Request a 32-bit depth buffer
```
Comment (jdilanzo) [inline Bottlebrush_Engine/Bottlebrush_Engine/BBScene/Scene.cpp:162]: This could go into the `Window` class.

**Review by jdilanzo — APPROVED** [review]

**Review by jdilanzo — APPROVED** [review]
Hella texture. Also make sure to merge `main` first.

**Review by Minywire — APPROVED** [review]

**Pull request merged**